### PR TITLE
updates for NNpack

### PR DESCRIPTION
**Pull request by CarloLucibello**
maxpool tests are currently failing due to some type conversions. We didn't catch this since the tests were added in #207 at the same time of #204 . 

Removing the type casts I got the tests passing, but then #210 convinced me to bypass NNPACK's maxpool entirely as it is not completely reliable. 
I also, added NNPACK's conv gradient as defaults, for some reason only the forward was used before (@dhairyagandhi96 @MikeInnes @avik-pal @staticfloat do you know why?)

So, this PR does the following.
- use NNPACK only for Float32 operations, fallback to NNlib for Float64
- use NNPACK for conv gradient  
- don't use NNPACK for maxpool as a workaround for #210. Notice that NNPACK doesn't offer a gradient for maxpool, so the damage is a little mitigated.

### CODE REVIEW

File: src/pooling.jl [hunk 1]
```diff
@@ -103,7 +103,7 @@ end
 
 
 # Finally, let's generate auto-allocating versions of all our functions, for all backends:
-for backend in (Symbol(), :_direct, :_im2col)
```
Comment (CarloLucibello) [inline src/pooling.jl:106]: the  im2col backend for pooling just doesn't exists

File: src/nnpack/interface.jl [hunk 2]
```diff
@@ -1,48 +1,30 @@
...
-
-
-function maxpool_nnpack!(y::Array{T1, 4}, x::Array{T2, 4}, pdims::PoolDims;
-                         kwargs...) where {T1, T2}
-    @warn "Automatically converting input tensor to Float32. This will have performance implications" maxlog=1
-    # We want the output to be of the same type as desired
-    T1.(maxpool_nnpack!(Float32.(y), Float32.(x), pdims; kwargs...))
-end
+## NNPACK supports only Float32
+## Uncomment if we decide to cast to Float32
```
Comment (avik-pal) [inline src/nnpack/interface.jl:4]: Not sure why we need to comment this out. If the user wants to use NNPACK with Float64 even after the warning I don't see any harm in typecasting and doing the operation. The presence of these functions shouldn't affect the exposed NNlib functions which should be using the fastest backend (which in this case would be the im2col/direct version)

Comment (CarloLucibello) [reply src/nnpack/interface.jl:4]: right, will change

File: src/conv.jl [hunk 3]
```diff
@@ -104,8 +104,9 @@ end
 # Finally, let's generate auto-allocating versions of all our functions, for all backends.
 # We `@timeit` these methods separately, as we want to know how much time is spent in
 # allocation.  :P
-for backend in (Symbol(), :_direct, :_im2col)
+for backend in (Symbol(), :_direct, :_im2col, :_nnpack)
     # First make auto-allocating versions of the conv()-like calls:
+    (backend == :_nnpack && !is_nnpack_available()) && continue
     for name in (:conv, :depthwiseconv)
```
Comment (avik-pal) [inline src/conv.jl:110]: I don't think we have `depthwiseconv_nnpack`

**Pull request merged**